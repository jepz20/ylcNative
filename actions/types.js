// @flow

export const DIGIT_PRESS = '@digit_press'
export const ERASE = '@erase'
export const SCOOP_MATCH = '@match/scoop_match'
export const HALF_POINTS = '@match/half_points'
export const ADD_POINTS = '@match/add_points'
export const DRAW = '@draw'
export const SUBSTRACT_POINTS = '@substract_points'
export const TOGGLE_CALCULATOR_VISIBILITY = '@toggle_calculator_visibility'
export const RESET_CALCULATOR_VALUE = '@rest_calculator_value'
export const CHANGE_PLAYER_NAME = '@change_player_name'

// Also setting the types
export type DIGIT_PRESS_STR = '@digit_press'
export type ERASE_STR = '@erase'
export type SCOOP_MATCH_STR = '@match/scoop_match'
export type HALF_POINTS_STR = '@match/half_points'
export type ADD_POINTS_STR = '@match/add_points'
export type DRAW_STR = '@draw'
export type SUBSTRACT_POINTS_STR = '@substract_points'
export type TOGGLE_CALCULATOR_VISIBILITY_STR = '@toggle_calculator_visibility'
export type RESET_CALCULATOR_VALUE_STR = '@rest_calculator_value'
export type CHANGE_PLAYER_NAME_STR = '@change_player_name'
