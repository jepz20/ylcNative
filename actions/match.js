// @flow

import * as T from './types'
import { getId } from '../utils'

type PlayerPayload = {
  player: string,
  logId: string
}

type PointsPayload = {
  points: string
}

type ScoopMatchAction = {
  type: T.SCOOP_MATCH_STR,
  payload: PlayerPayload
}

type DrawAction = {
  type: T.DRAW_STR,
  payload: { logId: string }
}

type HalfPointsAction = {
  type: T.HALF_POINTS_STR,
  payload: PlayerPayload
}

type AddPointsAction = {
  type: T.ADD_POINTS_STR,
  payload: PlayerPayload & PointsPayload
}

type SubstractPointsAction = {
  type: T.SUBSTRACT_POINTS_STR,
  payload: PlayerPayload & PointsPayload
}

export type Action =
  | ScoopMatchAction
  | DrawAction
  | HalfPointsAction
  | AddPointsAction
  | SubstractPointsAction

export type ScoopMatch = (player: string, logId: string) => ScoopMatchAction
export const scoopMatch: ScoopMatch = (player, logId) => ({
  type: T.SCOOP_MATCH,
  payload: { player, logId }
})

export type Draw = (logId: string) => DrawAction
export const draw: Draw = logId => ({
  type: T.DRAW,
  payload: {
    logId
  }
})

export type HalfPoints = (player: string, logId: string) => HalfPointsAction
export const halfPoints: HalfPoints = (player, logId) => ({
  type: T.HALF_POINTS,
  payload: { player, logId }
})

export type AddPoints = (
  points: string,
  player: string,
  logId: string
) => AddPointsAction
export const addPoints: AddPoints = (points, player, logId) => ({
  type: T.ADD_POINTS,
  payload: { points, player, logId }
})

export type SubstractPoints = (
  points: string,
  player: string,
  logId: string
) => SubstractPointsAction
export const substractPoints: SubstractPoints = (points, player, logId) => ({
  type: T.SUBSTRACT_POINTS,
  payload: { points, player, logId }
})
