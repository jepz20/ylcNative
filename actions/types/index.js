export const DIGIT_PRESS = '@digit_press'
export const ERASE = '@erase'
export const SCOOP_MATCH = '@scoop_match'
export const HALF_POINTS = '@half_points'
export const ADD_POINTS = '@add_points'
export const DRAW = '@draw'
export const SUBSTRACT_POINTS = '@substract_points'
export const SET_CURRENT_PLAYER = '@set_current_player'
export const TOGGLE_CALCULATOR_VISIBILITY = '@toggle_calculator_visibility'
export const RESET_CALCULATOR_VALUE = '@rest_calculator_value'
