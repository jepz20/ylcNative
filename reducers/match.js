// @flow

import { Record, Map, List } from 'immutable'
import type { RecordFactory } from 'immutable'
import {
  SCOOP_MATCH,
  DRAW,
  HALF_POINTS,
  ADD_POINTS,
  SUBSTRACT_POINTS
} from '../actions/types'
import type {
  StateRE,
  State,
  Result,
  Player,
  Log,
  LogList,
  LogMap,
  PlayerMap,
  ResultMap
} from '../types/match'
import type { Action } from '../actions/match'
import { parsePoints } from '../utils'

// TODO: Use AsyncStorage instead of hardcoded values
const DEFAULT_POINTS: number = 8000
const PLAYER_1_NAME: string = 'Jose'
const PLAYER_2_NAME: string = 'Eduardo'

const getId = (): string => Date.now().toString()

export const PlayerRecord: RecordFactory<Player> = Record({
  id: getId(),
  name: 'Default',
  currentPoints: DEFAULT_POINTS
})

export const LogsRecord: RecordFactory<Log> = Record({
  id: getId(),
  playerId: '0',
  operationValue: 0,
  currentPoints: 0,
  previousPoints: 0
})

export const resultsRecord: RecordFactory<Result> = Record({
  winner: null
})

const INITIAL_STATE: RecordFactory<State> = Record({
  players: Map({
    '1': PlayerRecord({
      id: '1',
      name: PLAYER_1_NAME,
      currentPoints: DEFAULT_POINTS
    }),
    '2': PlayerRecord({
      id: '2',
      name: PLAYER_2_NAME,
      currentPoints: DEFAULT_POINTS
    })
  }),
  currentDuel: '1',
  logs: Map(),
  results: Map()
})

const getPlayerPoints = (players: PlayerMap, player: string): number => {
  if (players.has(player)) {
    // $FlowFixMe
    return players.get(player).currentPoints
  }
  console.warn(players, player, 'getPlayerPoint')
  return 0
}

const addToLog = (
  currentDuel: string,
  logs: ?LogMap,
  playerId: string,
  operationValue: number,
  previousPoints: number,
  nextPoints: number
): LogMap => {
  const newLogs: LogMap = logs || Map()
  const newLog: LogList = (newLogs.get(currentDuel) || List()).push(
    LogsRecord({
      id: getId(),
      playerId,
      operationValue,
      currentPoints: nextPoints,
      previousPoints
    })
  )
  return newLogs.merge({ [currentDuel]: newLog })
}

const setDuelResult = (
  players: PlayerMap,
  currentDuel: string,
  results: ?ResultMap = Map()
): ResultMap => {
  const activePlayers: PlayerMap = players.filter(
    player => player.currentPoints > 0
  )

  const res: ResultMap = results || Map()
  const isWinner: boolean = activePlayers.size === 1
  const isTie: boolean = !activePlayers.size

  if (isWinner) {
    return res.merge({
      [currentDuel]: resultsRecord({
        // $FlowFixMe
        winner: activePlayers.first().id
      })
    })
  }

  if (isTie) {
    return res.merge({
      [currentDuel]: resultsRecord({ winner: 'tie' })
    })
  }

  return res
}

const operateValue = (
  state: StateRE,
  player: string,
  operationValue: number
): StateRE => {
  const previousPoints: number = getPlayerPoints(state.players, player)
  const nextPoints: number = previousPoints + operationValue

  const players = state.players.mergeDeep({
    [player]: {
      currentPoints: nextPoints
    }
  })
  const { currentDuel, logs, results } = state
  return state.merge({
    players: players,
    logs: addToLog(
      currentDuel,
      logs,
      player,
      operationValue,
      previousPoints,
      nextPoints
    ),
    results: setDuelResult(players, currentDuel, results)
  })
}

const draw = (state: StateRE): StateRE => {
  const players: PlayerMap = state.players.map(player =>
    player.set('currentPoints', 0)
  )
  return state.merge({
    players,
    results: setDuelResult(players, state.currentDuel, state.results)
  })
}

export default function (state: StateRE = INITIAL_STATE(), action: Action) {
  switch (action.type) {
    case ADD_POINTS: {
      const { points, player } = action.payload
      let operationValue: number = parsePoints(points)
      return operateValue(state, player, operationValue)
    }
    case SUBSTRACT_POINTS: {
      const { points, player } = action.payload
      let operationValue: number = parsePoints(points) * -1
      return operateValue(state, player, operationValue)
    }
    case HALF_POINTS: {
      const { player } = action.payload
      let operationValue: number =
        Math.ceil(getPlayerPoints(state.players, player) / 2) * -1
      return operateValue(state, player, operationValue)
    }
    case SCOOP_MATCH: {
      const { player } = action.payload
      let operationValue: number =
        Math.ceil(getPlayerPoints(state.players, player)) * -1
      return operateValue(state, player, operationValue)
    }
    case DRAW:
      return draw(state)
    default:
      return state
  }
}
