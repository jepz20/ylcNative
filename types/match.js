// @flow

import type { RecordOf, Map, List } from 'immutable'
export type Player = {
  id: string,
  name: string,
  currentPoints: number
}

export type Log = {
  id: string,
  playerId: string,
  operationValue: number,
  currentPoints: number,
  previousPoints: number,
  type: string
}

export type Result = {
  winner: ?string
}

export type PlayerRE = RecordOf<Player>
export type LogRE = RecordOf<Log>
export type ResultRE = RecordOf<Result>
export type LogList = List<LogRE>

export type PlayerMap = Map<string, PlayerRE>
export type LogMap = Map<string, LogList>
export type ResultMap = Map<string, ResultRE>

export type State = {
  players: PlayerMap,
  currentDuel: string,
  logs: LogMap,
  results: ResultMap
}

export type StateRE = RecordOf<State>
