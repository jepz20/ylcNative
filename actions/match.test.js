import * as A from './match.js'
import * as T from './types'
describe('Match Actions', () => {
  test('scoopMatch', () => {
    expect(A.scoopMatch('3')).toEqual({
      type: T.SCOOP_MATCH,
      payload: { player: '3' }
    })
  })

  test('draw', () => {
    expect(A.draw()).toEqual({
      type: T.DRAW
    })
  })

  test('halfPoints', () => {
    expect(A.halfPoints('3')).toEqual({
      type: T.HALF_POINTS,
      payload: { player: '3' }
    })
  })

  test('addPoints', () => {
    expect(A.addPoints('2210', '1')).toEqual({
      type: T.ADD_POINTS,
      payload: {
        points: '2210',
        player: '1'
      }
    })
  })

  test('substractPoints', () => {
    expect(A.substractPoints('1985', '2')).toEqual({
      type: T.SUBSTRACT_POINTS,
      payload: {
        points: '1985',
        player: '2'
      }
    })
  })
})
