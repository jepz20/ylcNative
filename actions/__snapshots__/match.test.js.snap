// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Match Actions substractPoints 1`] = `
Object {
  "payload": Object {
    "name": "test",
    "player": "2",
  },
  "type": "@change_player_name",
}
`;
